src/ray/common/ray_syncer/ray_syncer.h

```diff
@@ -0,0 +1,197 @@
+// Copyright 2022 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+#include <grpcpp/server.h>
+
+#include "absl/container/flat_hash_map.h"
+#include "absl/container/flat_hash_set.h"
+#include "boost/functional/hash.hpp"
+#include "ray/common/asio/instrumented_io_context.h"
+#include "ray/common/asio/periodical_runner.h"
+#include "ray/common/id.h"
+#include "src/ray/protobuf/ray_syncer.grpc.pb.h"
+
+namespace ray {
+namespace syncer {
+
+using ray::rpc::syncer::DummyRequest;
+using ray::rpc::syncer::DummyResponse;
+using ray::rpc::syncer::RayComponentId;
+using ray::rpc::syncer::RaySyncMessage;
+using ray::rpc::syncer::RaySyncMessages;
+using ray::rpc::syncer::StartSyncRequest;
+using ray::rpc::syncer::StartSyncResponse;
+
+static constexpr size_t kComponentArraySize =
+    static_cast<size_t>(ray::rpc::syncer::RayComponentId_ARRAYSIZE);
+
+/// The interface for a reporter. Reporter is defined to be a local module which would
+/// like to let the other nodes know its status. For example, local cluster resource
+/// manager.
+struct ReporterInterface {
+  /// Interface to get the snapshot of the component. It asks the module to take a
+  /// snapshot of the current status. Each snapshot is versioned, and it should return
+  /// std::nullopt if the version hasn't changed.
+  ///
+  /// \param current_version The version syncer module current has.
+  /// \param component_id The component id asked for.
+  virtual std::optional<RaySyncMessage> Snapshot(uint64_t current_version,
+                                                 RayComponentId component_id) const = 0;
+  virtual ~ReporterInterface() {}
+};
+
+/// The interface for a receiver. Receiver is defined to be a module which would like
+/// to get the status of other nodes. For example, cluster resource manager.
+struct ReceiverInterface {
+  /// Interface to update a module. The module should read the `sync_message` fields and
+  /// deserialize it to update its internal status.
+  ///
+  /// \param message The message received from remote node.
+  virtual void Update(std::shared_ptr<RaySyncMessage> message) = 0;
+  /// Indicate whether the message received by this module needs to be
+  /// broadcasted to other nodes.
+  virtual bool NeedBroadcast() const = 0;
+  virtual ~ReceiverInterface() {}
+};
+
+template <typename T>
+using Array = std::array<T, kComponentArraySize>;
```

Comment:
would a hash_map make the code cleaner?

Reply:
I feel hash map is overkill in this one. But anyway I don't think hash map will map the code cleaner, from my opinion, it seems ends with similar amount of code.

Reply:
Can we at least give a better name? I was confused about what Array is. Also seems directly using std::array<T, kComponentArraySize> is clearer?